File: src/op/dispatch/tvmjit/unary.cc
/*!
 * Copyright (c) 2019 by Contributors
 * \file ./src/op/dispatch/tvmjit/unary.cc
 * \brief Unary operators bridged from TVM.
 */
#include <array>
#include "./tvm_attrs.h"
#include "./tvmjit_utils.h"
#include "../../schema/ufunc.h"
#include "../../../common/shape_utils.h"

namespace mnm {
namespace op {
namespace tvmjit {

using namespace mnm::ir;
using common::shape_utils::GetNumel;
using schema::UnaryArgs;
using schema::UnaryDxArgs;
using schema::UnaryUfuncArgs;

std::vector<Value> UnarySchema2Args(const UnaryArgs* args) {
  return {args->x};
}

std::vector<std::string> UnarySchemaArgNames(const op::CallValues& call) {
  return {"x"};
}

MNM_TVMJIT(Copy, "mnm.op.copy", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Abs, "mnm.op.abs", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Ceil, "mnm.op.ceil", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Floor, "mnm.op.floor", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Log, "mnm.op.log", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Exp, "mnm.op.exp", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Cos, "mnm.op.cos", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Sin, "mnm.op.sin", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Sign, "mnm.op.sign", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Round, "mnm.op.round", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Relu, "mnm.op.relu", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Erf, "mnm.op.erf", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Sqrt, "mnm.op.sqrt", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Rsqrt, "mnm.op.rsqrt", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Atan, "mnm.op.atan", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(Negative, "mnm.op.negative", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames,
           GenericAttrs, GenericHasher);
MNM_TVMJIT(Sigmoid, "mnm.op.sigmoid", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames,
           GenericAttrs, GenericHasher);
MNM_TVMJIT(Tanh, "mnm.op.tanh", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);
MNM_TVMJIT(BatchFlatten, "mnm.op.batch_flatten", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames,
           GenericAttrs, GenericHasher);
MNM_TVMJIT(ZerosLike, "mnm.op.zeros_like", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames,
           GenericAttrs, GenericHasher);
MNM_TVMJIT(OnesLike, "mnm.op.ones_like", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames,
           GenericAttrs, GenericHasher);
MNM_TVMJIT(Trunc, "mnm.op.trunc", UnaryArgs, UnarySchema2Args, UnarySchemaArgNames, GenericAttrs,
           GenericHasher);

struct UnaryDxAttr : public tvm::AttrsNode<UnaryDxAttr> {
  std::string grad_mode;

  TVM_DECLARE_ATTRS(UnaryDxAttr, "relay.attrs.UnaryDxAttr") {
    TVM_ATTR_FIELD(grad_mode).describe(
        "Indicate how to calculate the gradient: using input, output or both");
  }
};

TVM_REGISTER_NODE_TYPE(UnaryDxAttr);

std::vector<Value> UnaryDxSchema2Args(const UnaryDxArgs* args) {
  CHECK(args->x.defined() || args->y.defined());
  std::vector<Value> ret;
  if (args->x.defined()) {
    ret.push_back(args->x.value());
  }
  if (args->y.defined()) {
    ret.push_back(args->y.value());
  }
  ret.push_back(args->dy);
  return ret;
}

std::vector<std::string> UnaryDxSchemaArgNames(const op::CallValues& call) {
  const auto* args = call->args.as<UnaryDxArgs>();
  CHECK(args->x.defined() || args->y.defined());
  std::vector<std::string> ret;
  if (args->x.defined()) {
    ret.push_back("x");
  }
  if (args->y.defined()) {
    ret.push_back("y");
  }
  ret.push_back("dy");

  return ret;
}

Attrs UnaryDxSchema2Attrs(const UnaryDxArgs* args) {
  auto attrs = make_object<UnaryDxAttr>();
  CHECK(args->x.defined() || args->y.defined());
  attrs->grad_mode = "both";
  if (!args->x.defined()) {
    attrs->grad_mode = "output";
  } else if (!args->y.defined()) {
    attrs->grad_mode = "input";
  }
  return Attrs(attrs);
}

MNM_TVMJIT_PLEVEL(ReluDx, "mnm.op.relu_dx", UnaryDxArgs, UnaryDxSchema2Args, UnaryDxSchemaArgNames,
                  UnaryDxSchema2Attrs, GenericHasher, 20);
MNM_TVMJIT(ErfDx, "mnm.op.erf_dx", UnaryDxArgs, UnaryDxSchema2Args, UnaryDxSchemaArgNames,
           UnaryDxSchema2Attrs, GenericHasher);
MNM_TVMJIT(TanhDx, "mnm.op.tanh_dx", UnaryDxArgs, UnaryDxSchema2Args, UnaryDxSchemaArgNames,
           UnaryDxSchema2Attrs, GenericHasher);
}  // namespace tvmjit
}  // namespace op
}  // namespace mnm

File: python/mnm/_tvm_op/unary.py
# pylint: disable=missing-function-docstring, unused-argument
"""Schedule registries for unary operators."""
import math
from .._lib import register_compute
from .._lib import tvm as _tvm
from .._lib import _reg


_reg.register_broadcast_schedule("mnm.op.tvm.copy")
_reg.register_broadcast_schedule("mnm.op.tvm.ceil")
_reg.register_broadcast_schedule("mnm.op.tvm.floor")
_reg.register_broadcast_schedule("mnm.op.tvm.abs")
_reg.register_broadcast_schedule("mnm.op.tvm.erf")
_reg.register_broadcast_schedule("mnm.op.tvm.cos")
_reg.register_broadcast_schedule("mnm.op.tvm.sin")
_reg.register_broadcast_schedule("mnm.op.tvm.sign")
_reg.register_broadcast_schedule("mnm.op.tvm.round")
_reg.register_broadcast_schedule("mnm.op.tvm.exp")
_reg.register_broadcast_schedule("mnm.op.tvm.log")
_reg.register_broadcast_schedule("mnm.op.tvm.log2")
_reg.register_broadcast_schedule("mnm.op.tvm.sqrt")
_reg.register_broadcast_schedule("mnm.op.tvm.rsqrt")
_reg.register_broadcast_schedule("mnm.op.tvm.atan")
_reg.register_broadcast_schedule("mnm.op.tvm.relu")
_reg.register_broadcast_schedule("mnm.op.tvm.negative")
_reg.register_broadcast_schedule("mnm.op.tvm.sigmoid")
_reg.register_broadcast_schedule("mnm.op.tvm.tanh")

def select_unary_dx_input(attrs, inputs, x_or_y):
    """Select the required input based on the grad_mode to calculate the gradient.
    x_or_y=True selects x; otherwise y.

    if grad_mode == "both":
        x, y, dy = inputs
    else:
        x_or_y, dy = inputs
    """
    grad_mode = attrs.grad_mode
    x_or_y_idx = 1 if not x_or_y and grad_mode == "both" else 0
    return inputs[x_or_y_idx], inputs[-1]


@register_compute("mnm.op.tvm.tanh_dx")
def tanh_dx_compute(attrs, inputs, output_type):
    # grad = dy * (1 - tanh(x) * tanh(x)) = dy * (1 - y * y)
    y, dy = select_unary_dx_input(attrs, inputs, False)
    return [_tvm.te.compute(y.shape, lambda *idx: dy[idx] * (1 - y[idx] * y[idx]),
                            tag=_tvm.topi.tag.ELEMWISE)]

_reg.register_broadcast_schedule("mnm.op.tvm.tanh_dx")
_reg.register_broadcast_schedule("mnm.op.tvm.trunc")

@register_compute("mnm.op.tvm.erf_dx")
def erf_dx_compute(attrs, inputs, output_type):
    x, dy = select_unary_dx_input(attrs, inputs, True)
    return [_tvm.te.compute(x.shape,
                            lambda *idx: _tvm.tir.const(2 / math.sqrt(math.pi), dtype=dy.dtype)
                            * _tvm.te.exp(-x[idx] * x[idx]) * dy[idx],
                            tag=_tvm.topi.tag.ELEMWISE)]

_reg.register_broadcast_schedule("mnm.op.tvm.erf_dx")
_reg.register_injective_schedule("mnm.op.tvm.zeros_like")
_reg.register_injective_schedule("mnm.op.tvm.ones_like")

@register_compute("mnm.op.tvm.sqrt_dx")
def sqrt_dx_compute(attrs, inputs, output_type):
    y, dy = select_unary_dx_input(attrs, inputs, False)
    return [_tvm.te.compute(y.shape,
                            lambda *idx: dy[idx] / (y[idx] + y[idx]),
                            tag=_tvm.topi.tag.ELEMWISE)]

_reg.register_injective_schedule("mnm.op.tvm.sqrt_dx")

@register_compute("mnm.op.tvm.gelu")
def gelu_compute(attrs, inputs, output_type):
    data = inputs[0]
    # gelu is data  * normcdf(data)
    const_point_5 = _tvm.tir.const(0.5, dtype=data.dtype)
    const_1 = _tvm.tir.const(1, dtype=data.dtype)
    const_sqrt_2 = _tvm.tir.const(math.sqrt(2), dtype=data.dtype)
    return [data * (const_point_5 * (const_1 + _tvm.topi.erf(data / const_sqrt_2)))]

_reg.register_injective_schedule("mnm.op.tvm.gelu")

@register_compute("mnm.op.tvm.gelu_dx")
def gelu_dx_compute(attrs, inputs, output_type):
    x, dy = select_unary_dx_input(attrs, inputs, True)
    const_point_5 = _tvm.tir.const(0.5, dtype=x.dtype)
    const_minus_point_5 = _tvm.tir.const(-0.5, dtype=x.dtype)
    const_1 = _tvm.tir.const(1, dtype=x.dtype)
    const_sqrt_2 = _tvm.tir.const(math.sqrt(2), dtype=x.dtype)
    const_sqrt_pi = _tvm.tir.const(math.sqrt(math.pi), dtype=x.dtype)
    # cdf = 0.5 * (1 + erf(x/sqrt(2)))
    cdf = const_point_5 * (const_1 + _tvm.topi.erf(x / const_sqrt_2))
    # beta = 1 / sqrt(2*pi)
    const_beta = const_1 / (const_sqrt_2 * const_sqrt_pi)
    # pdf = beta * e^(-0.5x^2)
    pdf = const_beta * _tvm.topi.exp(x * x * const_minus_point_5)
    return [dy * (cdf + x * pdf)]

_reg.register_injective_schedule("mnm.op.tvm.gelu_dx")
